### PR TITLE
feat(playground-v2): implement liquidity sources for playground

### PR DESCRIPTION
**Pull request by samobasquiat**

### CODE REVIEW

**Comment by mikasackermn** [comment]
[Screencast from 23-11-01 13:09:46.webm](https://github.com/rango-exchange/rango-client/assets/120931936/f1333b21-b223-453d-8db8-6a0c5e0b2ab9)
I think its behavior is strange for the first time !!

File: widget/playground/src/helpers.ts [hunk 1]
```diff
@@ -249,3 +249,19 @@ export function getCategoryNetworks(chains: BlockchainMeta[]) {
 
 export const tokenToString = (token: Asset) =>
   `${token.symbol}-${token.blockchain}-${token.address ?? ''}`;
```
Comment (mikasackermn) [inline widget/playground/src/helpers.ts:251]: can we move helpers to utils folder?

Comment (samobasquiat) [reply widget/playground/src/helpers.ts]: fixed.

**Review by mikasackermn — APPROVED** [review]

File: widget/playground/src/utils/common.ts [hunk 2]
```diff
@@ -17,3 +25,103 @@ export function shallowEqual<T>(
...
+  const seen = new Set<T[keyof T]>();
+  return arr.filter((item) => {
+    const identifier = item[key];
+    if (!seen.has(identifier)) {
+      seen.add(identifier);
+      return true;
+    }
+    return false;
+  });
+}
```
Comment (RanGojo) [inline widget/playground/src/utils/common.ts:46]: This is an AI suggestion :smiley_cat:  what do you think about that? I think your code is more readable but this one is more simplified. :thinking: 
```ts
export function removeDuplicates<T>(arr: T[], key?: keyof T): T[] {
  const seen = new Set<T[keyof T]>();

  return arr.filter(item => !key || typeof item[key] !== 'undefined' && !seen.has(item[key]) && seen.add(item[key]));
}
```

File: widget/playground/src/utils/common.ts [hunk 3]
```diff
@@ -17,3 +25,103 @@ export function shallowEqual<T>(
...
+  WalletTypes.LEAP,
+  WalletTypes.SAFE,
+  WalletTypes.MY_TON_WALLET,
+  WalletTypes.WALLET_CONNECT_2,
+  WalletTypes.BINANCE_CHAIN,
+];
+
+export const onChangeMultiSelects = (
+  value: string,
+  values: any[] | undefined,
```
Comment (RanGojo) [inline widget/playground/src/utils/common.ts:78]: I think we could move this function and the next one to another utils file :thinking:

File: widget/playground/src/utils/common.ts [hunk 4]
```diff
@@ -17,3 +25,103 @@ export function shallowEqual<T>(
...
+  const index = values.findIndex(findIndex);
+  if (index !== NOT_FOUND) {
+    values.splice(index, 1);
+  } else {
+    values.push(value);
+  }
+  if (values.length === list.length) {
+    return undefined;
+  }
+  return values;
```
Comment (RanGojo) [inline widget/playground/src/utils/common.ts:103]: `  return targetArray.length === list.length ? undefined : targetArray;
`

File: widget/playground/src/utils/common.ts [hunk 5]
```diff
@@ -17,3 +25,103 @@ export function shallowEqual<T>(
...
+  } else if (value === 'all') {
+    return undefined;
+  }
+  if (!values) {
+    values = [...list];
+    const index = list.findIndex(findIndex);
+    values.splice(index, 1);
+    return values;
+  }
+  values = [...values];
```
Comment (RanGojo) [inline widget/playground/src/utils/common.ts:93]: Is it same logic?

```ts
const modifiedValues = !values
  ? [...list].filter(findIndex)
  : [...values];
```

File: widget/playground/src/utils/export.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import type { WidgetConfig } from '@rango-dev/widget-embedded';
+
+import stringifyObject from 'stringify-object';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (RanGojo) [inline widget/playground/src/utils/export.ts:4]: I skip checking this file changes as it seems to be just moving functions to this file.

File: widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      ? []
+      : undefined
+    : isJustAllBridgeSelected
+    ? undefined
+    : selectedBridges;
+
+  return (
+    <>
+      <MultiSelect
+        label={`${isIncluded ? 'Exclude' : 'Supported'} DEXs`}
```
[the remark this answers is not in the document]
Comment (samobasquiat) [reply widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx:131]: fixed.

File: widget/embedded/src/types/config.ts [hunk 8]
```diff
@@ -136,4 +138,5 @@ export type WidgetConfig = {
   theme?: WidgetTheme;
   externalWallets?: boolean;
   pinnedTokens?: Asset[];
+  includeNewSources?: boolean;
```
[the remark this answers is not in the document]
Comment (samobasquiat) [reply widget/embedded/src/types/config.ts:141]: fixed.

File: widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        name: swapperGroup,
+      }));
+
+  const selectedDex = defaultSelectedItems('DEX');
+  const selectedBridges = defaultSelectedItems('BRIDGE');
+  const allExchanges = liquiditiesList('DEX');
+  const allBridges = liquiditiesList('BRIDGE');
+  const allExchangesNames = allExchanges.map((dex) => dex.name);
+  const allBridgeNames = allBridges.map((bridge) => bridge.name);
+  const isAllSelected = !liquiditySources;
```
[the remark this answers is not in the document]
Comment (samobasquiat) [reply widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx:54]: fixed.

File: widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        value={dexMultiValue}
+        defaultSelectedItems={
+          selectedDex || (isIncluded ? [] : allExchangesNames)
+        }
+        list={allExchanges}
+        onChange={handleDexChange}
+      />
+      <Divider size={12} />
+      <MultiSelect
+        label={`${isIncluded ? 'Exclude' : 'Supported'} Bridges`}
```
[the remark this answers is not in the document]
Comment (samobasquiat) [reply widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx:143]: fixed.

File: widget/playground/src/utils/common.ts (hunk 2 above)
Comment (samobasquiat) [reply widget/playground/src/utils/common.ts:46]: right.

File: widget/playground/src/utils/common.ts (hunk 3 above)
Comment (samobasquiat) [reply widget/playground/src/utils/common.ts:78]: onChangeMultiSelects should be removed at the end of implementation of the playground. we haven't removed unused and previous logics yet. But I transfered the next one.

File: widget/playground/src/utils/common.ts (hunk 4 above)
Comment (samobasquiat) [reply widget/playground/src/utils/common.ts:103]: As I mentioned above, this function will be removed.

File: widget/playground/src/utils/common.ts (hunk 5 above)
Comment (samobasquiat) [reply widget/playground/src/utils/common.ts:93]: Refer to the above comment.

File: widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+            (type === 'BRIDGE' && !swapper.types.includes('DEX')))
+      )
+    );
+
+  const liquiditiesList = (type: 'DEX' | 'BRIDGE') =>
+    uniqueSwappersGroup
+      .filter(
+        (swapper) =>
+          (type === 'DEX' && swapper.types.includes('DEX')) ||
+          (type === 'BRIDGE' && !swapper.types.includes('DEX'))
```
[the remark this answers is not in the document]
Comment (samobasquiat) [reply widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx:40]: fixed.

File: widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    ? undefined
+    : selectedDex;
+
+  const bridgeMultiValue = isAllSelected
+    ? isIncluded
+      ? []
+      : undefined
+    : isJustAllBridgeSelected
+    ? undefined
+    : selectedBridges;
```
[the remark this answers is not in the document]
Comment (samobasquiat) [reply widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx]: fixed.

File: widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    } else if (currentConfig.length === uniqueSwappersGroup.length) {
+      sources = isIncluded
+        ? removeDuplicates([...allBridgeNames, ...currentConfig])
+        : undefined;
+    } else {
+      sources = currentConfig;
+    }
+
+    onChangeSources(sources);
+  };
```
[the remark this answers is not in the document]
Comment (samobasquiat) [reply widget/playground/src/containers/FunctionalLayout/FunctionalLayout.Liquidities.tsx]: fixed.

**Pull request merged**